src/dex/dexalot/dexalot.ts

```diff
@@ -0,0 +1,935 @@
+
+    return null;
+  }
+
+  normalizeAddress(address: string): string {
+    return address.toLowerCase() === ETHER_ADDRESS
+      ? NULL_ADDRESS
+      : address.toLowerCase();
+  }
+
+  denormalizeAddress(address: string): string {
+    return address.toLowerCase() === NULL_ADDRESS
+      ? ETHER_ADDRESS
+      : address.toLowerCase();
+  }
+
+  // Dexalot protocol for native token expects 0x00000... instead of 0xeeeee...
+  normalizeToken(token: Token): Token {
+    return {
+      address: this.normalizeAddress(token.address),
+      decimals: token.decimals,
+    };
+  }
+
+  denormalizeToken(token: Token): Token {
+    return {
+      address: this.denormalizeAddress(token.address),
+      decimals: token.decimals,
+    };
+  }
+
+  calculateOrderPrice(
+    amounts: bigint[],
+    orderbook: string[][],
+    baseToken: Token,
+    quoteToken: Token,
+    side: ClobSide,
+  ) {
+    let result = [];
+
+    for (let i = 0; i < amounts.length; i++) {
+      let amt = amounts[i];
+      if (amt === 0n) {
+        result.push(amt);
+        continue;
+      }
+
+      let totalVolume = 0n;
+      let combinedPrice = 0n;
+
+      for (let j = 0; j < orderbook.length; j++) {
+        const order = orderbook[j];
+        const priceBN = ethers.utils.parseUnits(order[0], quoteToken.decimals);
+        const price = BigInt(priceBN.toString());
+        const qtyBN = ethers.utils.parseUnits(order[1], baseToken.decimals);
+        const qty = BigInt(qtyBN.toString());
+        let vol = qty;
+        if (side === ClobSide.BID) {
+          vol = (qty * price) / BigInt(10 ** baseToken.decimals);
+        }
+        if (amt < vol) {
+          totalVolume += amt;
+          combinedPrice += amt * price;
+          amt = 0n;
+        } else {
+          amt -= vol;
+          totalVolume += vol;
+          combinedPrice += vol * price;
+        }
+        if (amt === 0n) {
+          break;
+        }
+      }
+      if (amt > 0n) {
+        return result;
+      }
+      const avgPrice = combinedPrice / totalVolume;
+      if (side === ClobSide.BID) {
+        result.push((amounts[i] * BigInt(10 ** baseToken.decimals)) / avgPrice);
+      } else {
+        result.push((avgPrice * amounts[i]) / BigInt(10 ** baseToken.decimals));
+      }
+    }
+    return result;
+  }
+
+  async getPricesVolume(
+    srcToken: Token,
+    destToken: Token,
+    amounts: bigint[],
+    side: SwapSide,
+    blockNumber: number,
+    limitPools?: string[],
+  ): Promise<null | ExchangePrices<DexalotData>> {
+    try {
+      if (await this.isRestricted()) {
+        return null;
+      }
+
+      const normalizedSrcToken = this.normalizeToken(srcToken);
+      const normalizedDestToken = this.normalizeToken(destToken);
+
+      this.tokensMap = (await this.getCachedTokens()) || {};
+      if (normalizedSrcToken.address === normalizedDestToken.address) {
+        return null;
+      }
+
+      if (limitPools && limitPools.length !== 1) {
+        return null;
+      }
```

Comment:
I am not sure this the correct way of handling limitPools. We need to check the content, not only number of elements.

Reply:
@Verisana updated